[V1][TPU] Enable Top K

NickLucche opened the pull request:
Enabling the topk optimization that was introduced in https://github.com/vllm-project/vllm/pull/15242.

Currently facing the very issue foreseen by @njhill here https://github.com/vllm-project/vllm/pull/15242#issuecomment-2751743779.
```
ERROR 03-25 18:27:23 [core.py:343]     random_sampled = self.topk_topp_sampler(
ERROR 03-25 18:27:23 [core.py:343]                      ^^^^^^^^^^^^^^^^^^^^^^^
ERROR 03-25 18:27:23 [core.py:343]   File "/home/nick/vllm/.venv/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1751, in _wrapped_call_impl
ERROR 03-25 18:27:23 [core.py:343]     return self._call_impl(*args, **kwargs)
ERROR 03-25 18:27:23 [core.py:343]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 03-25 18:27:23 [core.py:343]   File "/home/nick/vllm/.venv/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1762, in _call_impl
ERROR 03-25 18:27:23 [core.py:343]     return forward_call(*args, **kwargs)
ERROR 03-25 18:27:23 [core.py:343]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 03-25 18:27:23 [core.py:343]   File "/home/nick/vllm/vllm/v1/sample/ops/topk_topp_sampler.py", line 119, in forward_tpu
ERROR 03-25 18:27:23 [core.py:343]     topk_values, topk_indices = torch.topk(logits, k, dim=-1)
ERROR 03-25 18:27:23 [core.py:343]                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 03-25 18:27:23 [core.py:343] TypeError: topk(): argument 'k' (position 2) must be int, not Tensor
```

~~Dumping work for reference, will look into it asap.~~

Update:
 - I've fixed topk only using @njhill work on https://github.com/vllm-project/vllm/pull/15478 which should land soon on main
 -  Removed the env variable that would fallback to native impl (as dicussed in prev pr)
 -  Addressed some of the cruft leftover from https://github.com/vllm-project/vllm/pull/15309


For completeness, I've run microbenchmarks and the new impl is slower (but of course correct):

```
//before
Running 32 elapsed time: 0.0018310546875
Running 32 elapsed time: 0.0017833709716796875
// after
 Running 32 elapsed time: 0.003275632858276367
Running 32 elapsed time: 0.003297090530395508
```
  
cc @hyeygit

NickLucche committed:
enable topk

Signed-off-by: NickLucche <nlucches@redhat.com>

NickLucche commented:
Let's hold until main is fixed to reduce entropy

hyeygit commented:
Thank you @NickLucche for this PR and thank you @njhill for fixing the batch case for top-k! In #15242 I only tested with the microbenchmark and test_sampler.py (scalar case only), not realizing that `k` can be batched. Thank you for the catch and sorry for the miss.

One thing to note is that on TPU `torch.topk` still involves a full vocab sort (see [XLA lowering](https://github.com/pytorch/xla/blob/79d1e86b7e99dc055f30f3a69aa71471c16f8dfe/torch_xla/csrc/xla_lower_util.cpp#L375)). The reason using `torch.topk` is so much faster on TPU is because of avoiding `torch.scatter` (as used in apply_top_k_top_p) which is extremely slow on TPU.

NickLucche commented:
The main blocker for this PR is topk recompilation. 
I was under the impression the XLA lowering would circumvent the need 
```python
import torch_xla.core.xla_model as xm
import torch_xla.debug.metrics as met
import torch

B, V = 3, 64
device = xm.xla_device()

logits = torch.randn(B, V, device=device)

# pre compile
k = 3
top_k_mask = logits.topk(k, dim=1).values
top_k_mask = top_k_mask.cpu()

print(met.short_metrics_report())
met.clear_all()

# Run
for k in [1, V]:
    top_k_mask = logits.topk(k, dim=1).values
    top_k_mask = top_k_mask.cpu()
    print(met.short_metrics_report()) # shows it compiles
    met.clear_all()
```

Is there something I am missing? @hyeygit @yaochengji

hyeygit committed:
[V1][TPU] TPU-optimized top-k implementation (2nd try)

Previously we found that using torch.topk resulted in significant
speed up for TPU. Turns out that's not a viable solution because the
return shape of torch.topk depends on k, which means an XLA recompilation
is triggered everytime k changes.

Additionally, we realized that torch.scatter was the main bottleneck for
the original top-k impl on TPU. This PR circumvents both problems by using
a threshold-based approach to find the top-k set. The algorithm is nearly
identical to that of top-p; see #15736 for more details.

Signed-off-by: Hyesoo Yang <hyeygit@gmail.com>

NickLucche commented:
@hyeygit I took the liberty of adding your contrib https://github.com/vllm-project/vllm/pull/15891 to this PR so we can test. Locally everything looks fine, let's wait for the CI tests. Nice work thanks!

NickLucche commented:
PS this PR should be merged after https://github.com/vllm-project/vllm/pull/15736 lands

NickLucche committed:
random k in test + fix rebase

Signed-off-by: NickLucche <nlucches@redhat.com>

NickLucche commented:
So enabling topk here is costing us a bit of performance. 
I am confident this is because the sampling graph got a bit bigger to execute and apart from the all_greedy batches, every request has to run through it.
Can you please confirm this on your side too? @yaochengji @hyeygit

I can propose further specializing the graph to separate topk/topp requests so that the rest is free to flow through (a separate graph), at the cost of increased code complexity. Let me know what you think.

hyeygit commented:
> So enabling topk here is costing us a bit of performance.

@NickLucche yes I think I can corroborate with this. I ran `benchmark_serving.py` with top-k and top-p enabled and observed a 6.8% drop in throughput (5.98 req/s -> 5.57 req/s).

alexm-redhat commented:
Just ran this PR for llama 70B with 8 x v6e TPUs. It achieves 5 reqs/sec instead of the previous 5.1 reqs/sec so the performance penalty I see is negligible.

yaochengji reviewed (approved):
LGTM! Thanks, Nick!

alexm-redhat commented on tests/v1/tpu/test_sampler.py:
```diff
@@ -39,3 +41,18 @@ def test_sampler_different(model_name: str):
+        sampling_params[0].top_k = -1
+        sampling_params[0].min_p = 0
+        output = llm.generate(p, sampling_params)
+        assert output[0].outputs[0].text != output[-1].outputs[0].text
```
I think the output of the first and last prompt should be different no matter if you set top_k to -1. Is it jus to see that it goes through?

maybe worth to have second and this prompts have same sampling_params and here just compare that outputs of 2 and 3 are the same (for some more sanity check).

NickLucche replied:
Thanks will address!

alexm-redhat reviewed (approved):
LGTM! Just some small comments

NickLucche committed:
simplify test

Signed-off-by: NickLucche <nlucches@redhat.com>

hyeygit commented:
@NickLucche small request -- could you incorporate the new top-k test from my PR? https://github.com/vllm-project/vllm/pull/15891/files#diff-09d15417fe42d494c51aaa9635ad51536751cc3c0659a7c4ce3b66bd6900eb1f

hyeygit commented on vllm/v1/sample/ops/topk_topp_sampler.py:
```diff
@@ -224,8 +227,8 @@ def apply_top_k_only(
-    k_index = k.sub_(1).unsqueeze(1).expand(logits.shape[0], 1)
-    top_k_mask = logits.topk(max_top_k, dim=1).values.gather(1, k_index.long())
+    k_index = k.sub_(1).unsqueeze(1)
+    top_k_mask = logits.topk(max_top_k, dim=1).values.gather(1, k_index)
```
Is this unintentional? This seems to be reverting what's at HEAD.

NickLucche replied:
yep this change was applied on top of @njhill original implementation when we first intended to use that function for TPU too.
Now that implementations have diverged, I am reverting to what we had before.

hyeygit replied:
I see. We might want to keep `k_index.long()` as that's not TPU-related. It was added in #15907.

NickLucche replied:
Right good catch, thanks for checking!

NickLucche committed:
topk equivalence tests

Co-authored-by: Hyesoo Yang <hyeygit@gmail.com>

Signed-off-by: NickLucche <nlucches@redhat.com>

NickLucche committed:
revert long

Signed-off-by: NickLucche <nlucches@redhat.com>

hgt312 commented:
default top_k pad value should not be zero? 0 trilgget the error in `probs_sort.gather(-1, top_k_count)`
